demos/slack-alert-app/slackbot.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2018- The Pixie Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+import logging
+import os
+import schedule
+import pxapi
+from slack.web.client import WebClient
+from slack.errors import SlackApiError
+
+
+# Pixie PxL script.
+PXL_SCRIPT = """
+import px
+
+df = px.DataFrame(table='http_events', start_time='-5m')
+
+# Add column for HTTP response status errors.
+df.error = df.http_resp_status >= 400
+
+# Add columns for service, namespace info
+df.namespace = df.ctx['namespace']
+df.service = df.ctx['service']
+
+# Filter for px-sock-shop namespace only.
+df = df[df.namespace == 'px-sock-shop']
+
+# Group HTTP events by service, counting errors and total HTTP events.
+df = df.groupby(['service']).agg(
+    error_count=('error', px.sum),
+    total_requests=('http_resp_status', px.count)
+)
+
+px.display(df, "status")
+"""
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+def get_pixie_data(cluster_conn):
+    # Get data from the Pixie API.
+
+    # Execute the PxL script.
+    script = cluster_conn.prepare_script(PXL_SCRIPT)
+    logging.debug("Pixie cluster executed script.")
+
+    service_stats_msg = ["*Recent 4xx+ Spikes in last 5 minutes:*"]
+
+    # Process table output rows to construct slack message.
+    for row in script.results("status"):
+        service_stats_msg.append(format_message(row["service"],
+                                               row["total_requests"],
+                                               row["error_count"]))
+
+    return "\n\n".join(service_stats_msg)
+
+
+def format_message(service, request_count, error_count):
+    # Format Pixie API table row data.
+    return (f"`{service}` \t ---> {error_count} (>4xx)"
+            f" errors out of {request_count} requests.")
+
+
+def send_message(slack_client, channel, cluster_conn):
+    # Send a POST request through the Slack Python client.
+
+    # Get data from the Pixie API.
+    msg = get_pixie_data(cluster_conn)
+
+    try:
+        logging.info(f"Sending {msg!r} to {channel!r}")
+        slack_client.chat_postMessage(channel=channel, text=msg)
+
+    except SlackApiError as e:
+        logging.error('Request to Slack API Failed: {}.'.format(e.response.status_code))
+        logging.error(e.response)
+
+
+if __name__ == "__main__":
+
+    # Get Pixie API key.
+    PIXIE_API_KEY = os.environ['PIXIE_API_KEY']
```

Comment:
might be good to put a check to see if the api key is defined in the environment

if 'PIXIE_API_KEY' not in os.environ:
     < some nice error msg >
     
same for the other env vars